tests/testsuite/cache_messages.rs

```diff
@@ -437,3 +386,50 @@ line 2
         )
         .run();
 }
+
+#[cargo_test]
+#[cfg(unix)]
+fn rustc_workspace_wrapper() {
+    use cargo_test_support::paths;
+
+    let p = project()
+        .file(
+            "src/lib.rs",
+            "pub fn f() { assert!(true); }\n\
+             fn unused_func() {}",
+        )
+        .build();
+
+    p.cargo("check -Zunstable-options -v")
+        .env("RUSTC_WORKSPACE_WRAPPER", paths::echo_wrapper().unwrap())
+        .masquerade_as_nightly_cargo()
+        .with_stdout_contains("WRAPPER CALLED: rustc --crate-name foo src/lib.rs [..]")
+        .run();
+
+    // Check without a wrapper should rebuild
+    p.cargo("check -v")
+        .with_stderr_contains(
+            "\
+[CHECKING] foo [..]
+[RUNNING] `rustc[..]
+[WARNING] [..]unused_func[..]
+",
+        )
+        .with_stdout_does_not_contain("WRAPPER CALLED: rustc --crate-name foo src/lib.rs [..]")
+        .run();
+
+    // Again, reading from the cache.
+    p.cargo("check -Zunstable-options -v")
+        .env("RUSTC_WORKSPACE_WRAPPER", paths::echo_wrapper().unwrap())
+        .masquerade_as_nightly_cargo()
+        .with_stderr_contains("[FRESH] foo [..]")
+        .with_stdout_does_not_contain("WRAPPER CALLED: rustc --crate-name foo src/lib.rs [..]")
```

Comment:
Sorry for the late reply.

The cache only caches stderr.  The wrapper script will need to echo to stderr.

I think I would be a little more comfortable if this ran on all platforms.  It might be easier to just build a rust program instead of using a shell script.  Can you maybe change it to something like this and remove the `#[cfg(unix)]` attributes?  It should take less than a second to build, so I'm not too worried about the impact.

```rust
pub fn echo_wrapper() -> std::path::PathBuf {
    let p = project()
        .at("rustc-echo-wrapper")
        .file("Cargo.toml", &basic_manifest("rustc-echo-wrapper", "1.0.0"))
        .file(
            "src/main.rs",
            r#"
            fn main() {
                let args = std::env::args().collect::<Vec<_>>();
                eprintln!("WRAPPER CALLED: {}", args[1..].join(" "));
                let status = std::process::Command::new(&args[1])
                    .args(&args[2..]).status().unwrap();
                std::process::exit(status.code().unwrap_or(1));
            }
            "#,
        )
        .build();
    p.cargo("build").run();
    p.bin("rustc-echo-wrapper")
}
```